Version module

bradcray opened the pull request:
This implements a new, standard, `Version` module with the goal of permitting Chapel programs to reason about version values in general, and the version of the Chapel compiler used to compile them specifically.  It takes the approach of using records with `param` fields to represent version numbers in order to support versions as a single logical value, yet one that supports compile-time comparisons.  New tests lock the behavior in.

In more detail, the module currently defines two version types:  `semanticVersion` representing a major.minor.update triple that encodes a semantic versioning style of number; and `sourceVersion` which is a semanticVersion coupled with a commit ID (e.g., a git SHA).  The components of these types are defined as `param` values in order to support compile-time comparisons of versions and code specialization.  Each of these types supports:

* being cast to a param string or getting printed out
* comparisons with other version values (of either type) via `==`, `!=`, `<`, `<=`, `>`, `>=`

The module also defines a `version()` convenience routine that takes major, minor, and optional update and commit values and returns a `semanticVersion` or `sourceVersion` value depending on whether or not the commit argument is supplied.  The concept is to support expressions like `if chplVersion > version(2,1) then...` and have the conditional be evaluated at compile-time.

Finally, it defines a `chplVersion` variable representing the version of the `chpl` compiler used to compile the program.

Possible future directions for this module:

* ability to convert strings into version values and/or compare version values to strings?
* queries that understand reasoning about semantic versioning like `.isCompatible()`, e.g., `if chplVersion.isCompatible(version(1,22)) then`...
* variables defining the source versions of the sources / modules that were used in compiling the program (where typically these will be identical to the `chplVersion`, but might be used to point to issues where the two don't match).

See the bottom of this post for screenshots of the chpldocumentation for this module.

As part of this effort, I also rejiggered the compiler to make it a bit more explicit when we're switching between an official release vs. a pre-release via a boolean flag in the version_num.h header.  I also converted the minor and update version numbers to integers for consistency and simplicity (and I couldn't see a reason that they would still need to be strings, if there ever was one.

Design questions:
- [ ] the current module is named `Version.chpl`... is that OK?
- [ ] the module supports `semanticVersion` and `sourceVersion` types... OK?
- [ ] the approach taken was to have a helper function named simply `version()` to provide an easy way for people to interact with these features (e.g., `if chplVersion > version(2,0)`), where the routine returns a value of one of the version types depending on the arguments.  This bucks the trend of calling the record `version` and the factory `createVersion()` or `version.create()` to optimize for the common case.  Is that OK?
- [ ] the `version()` routine currently only has default values for the update and commit arguments; @ronawho proposed that perhaps the minor version should default to 0 as well.  Thoughts?
- [ ] ~OK to both have `preRelease` (which seemed like the best way to phrase that boolean flag if it were a member) and `isOfficialRelease()` which seemed like the clearest way to support that query even though they're arguably redundant with one another?~

TODO:
- [ ] need to get review of names and values of variables supported
- [ ] need to alert RAD team to a simplification for what needs to change in version.cpp around release time

Resolves #5491.

<img width="664" alt="Screen Shot 2020-06-10 at 4 41 11 PM" src="https://user-images.githubusercontent.com/7536222/84329362-65f10000-ab39-11ea-9d05-cee122d8bd69.png">
<img width="667" alt="Screen Shot 2020-06-10 at 4 40 57 PM" src="https://user-images.githubusercontent.com/7536222/84329376-6e493b00-ab39-11ea-95d6-873e6ae62bde.png">
<img width="639" alt="Screen Shot 2020-06-10 at 4 41 19 PM" src="https://user-images.githubusercontent.com/7536222/84329390-730def00-ab39-11ea-9a24-c30acf59c271.png">

mppf commented on compiler/main/version.cpp:
```diff
@@ -68,3 +76,38 @@ get_clang_sysroot_args() {
+  printf("/* An ``int`` representing the update version number of `chpl` */\n");
+  printf("  param CHPL_VERSION_UPDATE: int = %s;\n", UPDATE_VERSION);
+  printf("/* A ``bool`` indicating whether or not this is an official release of `chpl` */\n");
+  printf("  param CHPL_VERSION_OFFICIAL: bool = ");
```
I'd probably call this CHPL_VERSION_IS_RELEASE.

bradcray replied:
Based on the other feedback on issue #5491, I ended up changing it to `versionIsRelease`.

mppf commented:
I'm a bit surprised that this took the approach of generating a .chpl file rather than having the compiler creates some params - the way that e.g. gBoundsChecking / boundsChecking works. I'd make them have names like chpl_CHPL_VERSION and then the ChapelVersion file would just export better names for these that are supported by the compiler.

Oh yeah, and I tend to prefer ChapelVersion vs ChplVersion for the name of the module.

Generating a .chpl file seems more error-prone (what if the version changes in-between? Do we have to include it in releases? What if the compiler is using the wrong modules path? What if the path to the generated file doesn't include some tidbit that is important, such as whether the compiler was built with LLVM?).

bradcray commented:
> I'm a bit surprised that this took the approach of generating a .chpl file rather than having the compiler creates some params - the way that e.g. gBoundsChecking / boundsChecking works.

I think some downsides to having the compiler generate params is that (a) they tend to be harder to document (e.g., I'm not seeing any documentation on `boundsChecking`), and (b) there's no way to opt out of `use`ing them (though that's probably not a huge deal).  It also somehow feels more special and subtle to me than embedding the values clearly in user-readable code.  I modeled this choice after ChapelEnv.chpl which could arguably also be compiler-generated.

> Generating a .chpl file seems more error-prone

I agree there are tradeoffs here...

> what if the version changes in-between?

In-between what?  (Each build of the compiler or docs (from the top-level directory at least) should re-generate the file).

> Do we have to include it in releases?

That seems like a good idea, yes.

> What if the compiler is using the wrong modules path?

That would be a bug.

> What if the path to the generated file doesn't include some tidbit that is important, such as whether the compiler was built with LLVM?).

My approach was to have the path to the module mirror the path to the compiler under `bin/`.  Is there a reason that such tidbits would be important in the documentation case and not in the path-to-`chpl` case?

mppf commented:
> I agree there are tradeoffs here...

It is OK to do it this way if it seems much better.

> I think some downsides to having the compiler generate params is that (a) they tend to be harder to document (e.g., I'm not seeing any documentation on boundsChecking), 

I'd think we'd document such things in `modules/internal/UtilMisc_forDocs.chpl`.

> and (b) there's no way to opt out of useing them (though that's probably not a huge deal).

I don't see a problem with that - the compiler would use e.g. `chpl_SomeVariable` as the param it introduces, and then the module you make would have proper names that respect scoping.

> In-between what? (Each build of the compiler or docs (from the top-level directory at least) should re-generate the file).

In-between updating&rebuilding the compiler and running `chpl`. I think you have more faith in `make` than I do...

>  Is there a reason that such tidbits would be important in the documentation case and not in the path-to-chpl case?

I don't have any specific concern here - I was just imagining potential problems.

ronawho commented:
>  the current module is named ChapelVersion.chpl... should it simply be Version.chpl?

I might have Version.chpl that contains the record, and ChapelVersion that uses it (I can imagine version being useful for other package version comparisons)

>  in code, should we use sha or SHA?

I'd probably lean towards SHA, but I also wonder if we should use something more generic if say in the future we switch away from git (seems a little silly to think about, but when the project was started we probably didn't imagine switching off cvs/svn)

> the approach taken was to have a helper function named simply version() because we think that's how people will interact with these features the most (e.g., if chplVersion > version(2,0)), though this results in calling the underlying type versionValue. This bucks the trend of calling the record version and the factory createVersion() or version.create() to optimize for the common case. Is that acceptable to people?

I like having just `version()`, I think that looks much cleaner.

> OK to both have preRelease (which seemed like the best way to phrase that boolean flag if it were a member) and isOfficialRelease() which seemed like the clearest way to support that query even though they're arguably redundant with one another?

If we keep both, I'd probably try to unify the names (both or neither starts with `is`). Do you have an example for when you'd expect either preRelease or officialRelease to be used? 

---

Another question that came to mind is if it'd be worth being able to create a version from a string. `version.parse("1.22.0")` or whatever

ronawho commented on modules/standard/ChapelVersion.chpl:
```diff
@@ -0,0 +1,230 @@
+    :returns: A new version value
+  */
+     
+  proc version(param major: int, param minor: int,
```
Any reason to not have `minor` default to 0 as well?

bradcray replied:
There's no real reason, no.  I was thinking in practice any reasonable query of a version number would want to compare at least major and minor and that `version(2,0)` said that more clearly than `version(2)`.  I worry a little that `version(2)` might suggest to someone `2.*.*` (any version in the 2.x.y family) rather than `2.0.0`.

bradcray replied:
I've captured this question in the OP as a design question to converge on.

ben-albrecht commented:
My 2₵:

> the current module is named ChapelVersion.chpl... should it simply be Version.chpl?

I prefer `ChapelVersion`. It is immediately clear to me what it does in any context. A user might want to implement a `Version` submodule in their package.

> in code, should we use sha or SHA?

I prefer `sha` - it's camelCase and requires less effort to write. Other ideas include `hash` and `checksum` (`checkSum`?), but I like `sha` the most.

> the approach taken was to have a helper function named simply version() because we think that's how people will interact with these features the most (e.g., if chplVersion > version(2,0)), though this results in calling the underlying type versionValue. This bucks the trend of calling the record version and the factory createVersion() or version.create() to optimize for the common case. Is that acceptable to people?

I am on the fence about this, but it's probably OK.

> OK to both have preRelease (which seemed like the best way to phrase that boolean flag if it were a member) and isOfficialRelease() which seemed like the clearest way to support that query even though they're arguably redundant with one another?

I am not sure the redundancy is worth it. I'd prefer only one - probably `preRelease`  since it's shorter.

ronawho commented on modules/standard/ChapelVersion.chpl:
```diff
@@ -0,0 +1,230 @@
+    }
+  }
+
+  proc ==(v1: versionValue(?), v2: versionValue(?)) param {
```
There's a lot of very similar code for all these comparisons. Looking at the python distutils version implementation [here](https://docs.openstack.org/tooz/latest/_modules/distutils/version.html), I wonder if we could implement a single function that returned `-1 for <`, `0 for =`, and `1 for >` and use that to implement the comparison methods.

On the other hand I suppose what's here is already implemented, so maybe it's not worth the effort.

bradcray replied:
I was thinking that the `compilerError()` would be a challenge to that "spaceship operator"-like approach, but I think that was when I still had some errors in my logic.  I also wouldn't want to call the helper twice for `<` and `=` (which might suggest returning -2, -1, 0, 1, and 2 or something like that?).  Mostly, it was easiest for me to think in these terms once I got the first one correct.

mppf replied:
Re. "spaceship operator" - one day, I hope we will have the `<==>` operator that can implement all of these at once (see #8545)

bradcray replied:
In a subsequent draft, I used a private routine named `spaceship` to avoid a lot of the repeated code in other cases.  Micro-optimizing, it's interesting to me to see how some queries might be answered more quickly/directly by explicit operator overloads than by a general spaceship operator, though for this case it's not likely to be a factor (given that these are evaluated at compile-time).

e-kayrakli commented:
> the current module is named ChapelVersion.chpl... should it simply be Version.chpl?

+1 on what Elliot and Ben said

> in code, should we use sha or SHA?

No preference

> the approach taken was to have a helper function named simply version() because we think that's how people will interact with these features the most (e.g., if chplVersion > version(2,0)), though this results in calling the underlying type versionValue. This bucks the trend of calling the record version and the factory createVersion() or version.create() to optimize for the common case. Is that acceptable to people?

I'm OK with it.

> OK to both have preRelease (which seemed like the best way to phrase that boolean flag if it were a member) and isOfficialRelease() which seemed like the clearest way to support that query even though they're arguably redundant with one another?

+1 on what Ben said: I prefer having a single flag for this, and I'd prefer `preRelease`. It seems to be the "exceptional" case between pre-release and official release for the uses of this module

> Another question that came to mind is if it'd be worth being able to create a version from a string. version.parse("1.22.0") or whatever

... and comparing a version with a string maybe? Or is it too much? `chplVersion == "1.22.0"` looks cute.

mppf commented:
I really like @ronawho's suggestion here, FWIW

> I might have Version.chpl that contains the record, and ChapelVersion that uses it (I can imagine version being useful for other package version comparisons)

bradcray commented:
Replying to @ronawho's comments:

> I might have Version.chpl that contains the record, and ChapelVersion that uses it (I can imagine version being useful for other package version comparisons)

I agree that this module is potentially more broadly applcable than to just the version of `chpl` (e.g., I wanted to also have a `srcVersion` variable that indicated the version of the sources in CHPL_HOME so that a warning could be printed if `chplVersion != srcVersion`) which is partially what made me think we should just call this Version.chpl.  What do you see the benefit being in having `ChapelVersion` that only defines `chplVersion` rather than just having a variable named `chplVersion` within `Version.chpl`?

> but I also wonder if we should use something more generic if say in the future we switch away from git 

I'm open to that given suggestions.  Note that SHA is a general term that's not particularly specific to git (but that doesn't guarantee that git's replacement will use it of course...).

> Do you have an example for when you'd expect either preRelease or officialRelease to be used?

I expected most users to use `isOfficialRelease()` but I wanted the ability to introduce a `param bool` to store this value in the future which is where `preRelease` came from.  I think of `preRelease` as being the property of the version value and `isOfficialRelease()` as being to user-facing query to get at it.  But that perhaps argues that `preRelease` should be private or undocumented until it actually is a field.

mppf commented:
> I'm open to that given suggestions. Note that SHA is a general term that's not particularly specific to git (but that doesn't guarantee that git's replacement will use it of course...).

Could it be simply called `commit` in the code? After all, that's the purpose of the `sha`, right? If we were using SVN wouldn't it be reasonable to have a number like `r1768915` there?

bradcray commented:
w.r.t. suggestions about string utility functions, since this PR has already been open a year (wow, exactly a year as of today?!?  Or does GitHub have a bug in it?), I'm going to propose we focus on getting it into a shape that it's acceptable to people and then add more features to it later rather than trying to add everything to it now that we might someday want (I backed away from my `srcVersion` idea above for similar reasons).

ronawho commented:
> What do you see the benefit being in having ChapelVersion that only defines chplVersion rather than just having a variable named chplVersion within Version.chpl?

I'm not sure there's a concrete benefit to having a separate ChapelVersion, just seems distinct from a general version mechanism so it made sense to me to split them. What came to mind was "if I just want to compare versions for my project, why is that bringing in chplVersion"?  If I had to rank my choices I would prefer:
 - separate Version and ChapelVersion modules
 - Only a Version module that had chplVersion
 - Only a ChapelVersion modul

bradcray commented:
Focusing in on the ChapelVersion vs. Version issue, I think Elliot's enumeration of the three options is the right one, though I'll also add a fourth:

1. Sibling `Version` and `ChapelVersion` modules
2. A single module, `Version`, that includes `chplVersion`
3. A single module, but named `ChapelVersion`
4. A `Version` module with submodules like `Compiler` or `Language` or `Source`, each of which focuses on representing the version of that thing (where, for this PR, we'd only have `Compiler` and it would only store `chplVersion`).

Option 2 vs. Option 1
---------------------
I lean pretty strongly towards option 2 for the sake of:

* simplicity (a single module rather than two)
* minimalism (I'd rather have fewer standard modules with more general names/capabilities than more standard modules with very specific roles); I think some modules have problems with sprawl, but this one doesn't seem likely to have that issue
* avoiding the case of having modules that only define a single symbol (which feels like overkill to me)

Michael and Elliot prefer option 1.  Elliot explains:

> What came to mind was "if I just want to compare versions for my project, why is that bringing in chplVersion"?

I expect that programmers who are concerned about too many symbols coming in from a module would also be ones who would rely on `import` (or filtered `use`) over `use` and so wouldn't have anything polluting their namespace that they didn't request or want.

Option 3
---------
I'm pretty against about option 3 because I think version types and capabilities are valuable for things other than querying the compiler's version number.  So to me, adding `Chapel` to the name would be a bit like calling our modules `ChapelMath` or `ChapelIO` rather than `Math` and `IO`.  

BenA argues for option 3 for the following reasons:

> It is immediately clear to me what it does in any context. 

But that seems to suggest that he thinks it's doing something other than what I do (i.e., not providing general capabilities for reasoning about versions, just focused on Chapel's version).

> A user might want to implement a Version submodule in their package.

I don't think naming it `Version` prevents them from doing that, and `import` has been designed to be very unambiguous w.r.t. which is being referenced (I'd argue `use` is also unambiguous, though it's more about understanding the rules than making the source code explicit).  (That said, this comment also brings to mind Bryant's #12712 which would provide additional cover for simply naming it `Version`).


Option 4
---------

I'm not a big fan of this option for similar reasons as option 1, but I prefer it over option 1 because it keeps version-related stuff together.  I'm just not excited about having to `import Version.CompilerVersion.chplVersion;` in order to write `if chplVersion > version(1.22.0)`.


Misc
-----

* Engin says:

  > +1 on what Elliot and Ben said

  But they were saying different things, so I'm not sure whether he prefers option 1 or 3

* Writing this up has made me realize a few things:

  * if `versionValue` is designed to be used for things other than the compilerVersion, it should print itself out more generically
  * it might make sense to make the current `versionValue` type into `gitVersion` or `sourceVersion`; and then to have a more general `semanticVersion` type that only stored `major.minor.update` (where the former would include the latter).

  So I'll add additional design questions at the top for these

e-kayrakli commented:
> But they were saying different things, so I'm not sure whether he prefers option 1 or 3

Sorry for being vague. I was mostly against 2. I can't really explain why but `Version` sounds too general to me and have concerns similar to Ben's. Maybe if it was named `Versioning` or `VersionSupport` or something, I would think differently?

ben-albrecht commented:
> BenA argues for option 3 for the following reasons:
> 
>> It is immediately clear to me what it does in any context.
> 
> But that seems to suggest that he thinks it's doing something other than what I do (i.e., not providing general capabilities for reasoning about versions, just focused on Chapel's version).

That's correct. `Version.chplVersion` or `Version.compiler` is also immediately clear to me. I misinterpreted the proposal as `Version.version` being the Chapel version.

>> A user might want to implement a Version submodule in their package.
> 
> I don't think naming it Version prevents them from doing that, and import has been designed to be very unambiguous w.r.t. which is being referenced (I'd argue use is also unambiguous, though it's more about understanding the rules than making the source code explicit). (That said, this comment also brings to mind Bryant's #12712 which would provide additional cover for simply naming it Version).

That's a good point. Also, if `Version` is a standardized way to deal with version numbers, it would be preferable the users rely on it vs. implement their own `Version` submodule.

Now that I better understand the proposal (or that the proposal has evolved), I am not a fan of (3) for similar reasons @bradcray points out. I'm not sure where I stand on (1), (2), and (4). They all seem like reasonable options at this point.

mppf commented:
> Michael and Elliot prefer option 1. Elliot explains:

To be clear, it's mostly that I don't like Option 3. Reading the above, Option 2 above appeals the most to me. For a second I thought "Ooh, 4 lets us have a Chapel Language version" but in Option 2 we'd just have something like `Version.languageVersion`.

> the approach taken was to have a helper function named simply version() because we think that's how people will interact with these features the most (e.g., if chplVersion > version(2,0)), though this results in calling the underlying type versionValue. This bucks the trend of calling the record version and the factory createVersion() or version.create() to optimize for the common case. Is that acceptable to people?

I suppose I'd lean towards naming the record `version` and just writing e.g. `if chplVersion > new version(2,0))` but I think the current approach is acceptable. If that seemed too verbose, I'd simply add some tuple comparison overloads, so you could write `if chplVersion > (2,0)`. Either way, we are expecting functions returning versions of a particular thing like `chplVersion` to indicate which thing in their name.

bradcray commented:
I took another pass on this based on feedback and my evolution in thinking based on the feedback (though there is other feedback still to be addressed).  Most notably, I:

* split `versionValue` into two types: `semanticVersion` and `gitVersion` to reflect `major.minor.update` vs. `major.minor.update+SHA`  I support comparisons between any two values of both types
* I liked Michael's idea of `commit` rather than `sha` for the field name in `gitVersion`, but then backed away from it because an `svnVersion` value could be given a complete ordering whereas a `gitVersion` value can't (or at least, can't at easily at compile-time).  So I basically doubled down on making that version type git-specific rather than having a more general `sourceVersion` or `rcsVersion` or something like that.
* I made `chplVersion` and the `version()` utility routine both vary in type: `chplVersion` will be a `semanticVersion` for official releases and a `gitVersion` for pre-releases.  This is a little bit funny, but I haven't come up with a case where it actually causes problems yet.  Similarly, the `version()` helper will return a `semanticVersion` if not provided a SHA (or provided an empty one), and a `gitVersion` otherwise.
* I punted on the module naming question for now to focus more on its guts, but will come back to that
* I ended up creating a pair of spaceship-like things because once I had to add a bunch more overloads of comparison operators, the code explosion was getting out of hand.

Here are the current docs:

<img width="666" alt="Screen Shot 2020-06-09 at 6 01 29 PM" src="https://user-images.githubusercontent.com/7536222/84215803-53fd5780-aa7c-11ea-8f92-e833e16e1a20.png">
<img width="670" alt="Screen Shot 2020-06-09 at 6 01 37 PM" src="https://user-images.githubusercontent.com/7536222/84215809-58297500-aa7c-11ea-8a6c-64bd1bda02e8.png">
<img width="680" alt="Screen Shot 2020-06-09 at 6 01 43 PM" src="https://user-images.githubusercontent.com/7536222/84215812-5bbcfc00-aa7c-11ea-84a4-faa0de7e2cf1.png">
<img width="674" alt="Screen Shot 2020-06-09 at 6 01 49 PM" src="https://user-images.githubusercontent.com/7536222/84215814-5f508300-aa7c-11ea-8cc4-204567bfe9fd.png">

And then there are a ton of comparison operator signatures as well that I didn't bother snapshotting...

bradcray commented:
In its current form, the PR and OP describing it implements my best attempt at addressing the comments so far and is worth checking in on again.  I've put the latest version of the rendered chpldocs for the module into the OP for reference.

mppf commented on modules/standard/Version.chpl:
```diff
@@ -0,0 +1,349 @@
+    forwarding version only major, minor, update;
+
+    // TODO: Why does the following type declaration break compilation?
+    //    const version /*: semanticVersion(?,?,?) */;
```
Perhaps you have to use just one `?`

bradcray replied:
No, that gave the same behavior.

bradcray replied:
Split this off into issue #15830.

mppf commented on modules/standard/Version.chpl:
```diff
@@ -0,0 +1,349 @@
+
+    Note that comparisons between two sourceVersion values with identical
+    semantic versions are not supported due to the challenges involved
+    in ordering commit values.  In addition, when a semanticVersion is
```
I think of `==` and `!=` as comparisons but these are implemented. Maybe say "comparisons other than `==` and `!=`"

bradcray replied:
Good point.  Does using the term "ordered comparisons" fix this for you?

mppf replied:
I think that would be good enough, yes

ronawho commented on modules/standard/Version.chpl:
```diff
@@ -0,0 +1,349 @@
+* compile-time comparisons between version values via ``==``, ``!=``,
+  ``<``, ``<=``, ``>``, and ``>``, including mixed types.  Generally
+  speaking, "less than" corresponds to "is an earlier version than."
+
```
It'd be nice to have a code example here showing how to compare something like `chplVersion > version(1, 22)`, but I'm struggling to come up with an example that isn't too artificial and won't become out of date each release.

bradcray replied:
What do you think about:

```chapel
    if chplVersion < version(2,0) then                                          
      writeln("Warning: Using a not-yet-stable version of Chapel");             
```

bradcray replied:
Or maybe instead of `writeln()`, `compilerWarning()` to subtly emphasize that it's a compile-time folded conditional?

ronawho replied:
Yeah, I like the compiler warning

bradcray replied:
Thinking about this more over lunch, an alternative that might be less revealing of Chapel's stability would be something like:

```chapel
if chplVersion < version(1,22) then
  compilerError("Arkouda is intended to be used with Chapel 1.22 or later");
```

(of course, this conditional by definition will never be false, but we don't have to admit to that... :) ).

The only slightly weird thing about this is that it mentions Arkouda in our docs.  Could say something vaguer like "This code..." instead if that seemed weird.

ronawho commented on modules/standard/Version.chpl:
```diff
@@ -0,0 +1,349 @@
+              returns a value of :type:`semanticVersion` otherwise it
+              returns a value of :type:`sourceVersion`.
+  */
+  proc version(param major: int, param minor: int,
```
I'd still be in favor of giving `minor` a default value so down the road you can do things like: `if chplVersion > version(2)`

ronawho replied:
Whoops, I had missed your original response in https://github.com/chapel-lang/chapel/pull/13191#discussion_r436162130. I can see that argument, but I was thinking that there's a lot of times when I'm comparing gcc versions where all I care about is the major version. That said, having to write `if chplVersion > version(2, 0)` isn't much more work, so maybe that's not worthwhile.

ronawho commented on modules/standard/Version.chpl:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  proc <(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
+    return spaceship(v1,v2) == -1;
```
I personally find these easier to think about/parse when written in terms of comparisons to 0:

 - `proc ==` -> `== 0`
 - `proc < `-> `< 0`
 - `proc <=` -> `<= 0`

I'm not sure it's worth changing since this is tested, I just have a slow time looking at these and convincing myself they're right.

bradcray replied:
Good suggestion.  I'm not familiar enough with the spaceship operator to have thought of this pattern.

bradcray commented:
I spawned the design questions on this PR off to #15911 just because (a) it's probably important not to break things after 1.23 and (b) I'm not sure how many eyes they got (I didn't hear much specific to them other than the original questions that led to them).
